### PR TITLE
Repair some corner cases in cancellation propagation between coroutines and listenable futures 

### PR DESCRIPTION
**Pull request by vadimsemenov**
Implement bidirectional cancellation for `future` coroutine builder.

This also:

* Refactors JobListenableFuture infrastructure so it can be reused in CoroutineScope.future and Deferred.asListenableFuture;
* Provides more descriptive `toString` implementation for the returned Future;
* Fixes stack traces in thrown exception, so it includes call to get() that triggered the exception to be thrown;
* Hides ListenableFuture.asDeferred return type, so it can't be casted to CompletableDeferred;
* Adds more tests to cover fixed corner cases;
* Improves documentation;
* Suppresses annoying warnings in tests.

### CODE REVIEW

**Comment by vadimsemenov** [comment]
Fixes #1442

**Comment by vadimsemenov** [comment]
Ping?

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 1]
```diff
@@ -271,57 +278,46 @@ private class ToContinuation<T>(
...
-     * the [Deferred] and [ListenableFuture] types, while permitting both kinds of promise to point
-     * to the same running task.
-     */
-    override fun run() {
-        if (future.isCancelled) {
-            cancel()
-        }
-    }
+    // JobListenableFuture propagates external cancellation to `this` coroutine. See JobListenableFuture.
+    val future = JobListenableFuture<T>(this)
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:305]: `@JvmField` will save an accessor here (which is redundant)

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 2]
```diff
@@ -271,57 +278,46 @@ private class ToContinuation<T>(
...
     override fun onCompleted(value: T) {
-        future.set(value)
+        future.complete(value)
     }
 
-    // TODO: This doesn't actually cancel the Future. There doesn't seem to be bidi cancellation?
     override fun onCancelled(cause: Throwable, handled: Boolean) {
-        if (!future.setException(cause) && !handled) {
-            // prevents loss of exception that was not handled by parent & could not be set to SettableFuture
+        if (!future.completeExceptionally(cause) && !handled) {
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:312]: It seems like (a long-existing) problem here: if `future {}` is cancelled **normally** (e.g. its parent got cancelled) and meanwhile someone completes the future from the outside, then `JobCancellationException` got reported into an exception handler, while we guarantee that no `CancellationException` will ever reach it.

This problem seems to exist in most of our `AbstractCoroutine` implementors, unfortunately. It would be nice if you could write a test that reproduces the problem and a fix for that (see: `MonoCoroutine.onCancelled`), I'll take care of the rest of the builders

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 3]
```diff
@@ -333,130 +329,161 @@ private class ListenableFutureCoroutine<T>(
...
+ *   - To maintain sanity, this class implements [ListenableFuture] and uses an auxiliary [SettableFuture]
+ *     around coroutine's result as a state engine to establish happens-after-completion. This
+ *     could probably be compressed into one subclass of [AbstractFuture] to save an allocation, at the
  *     cost of the implementation's readability.
  */
-private class OuterFuture<T>(private val deferred: Deferred<T>): ListenableFuture<T> {
-    val innerFuture = DeferredListenableFuture(deferred)
+private class JobListenableFuture<T>(private val jobToCancel: Job): ListenableFuture<T> {
+    /**
+     * Serves as a state machine for [Future] cancellation.
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:340]: Great job with auxulary comments 👍

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 4]
```diff
@@ -333,130 +329,161 @@ private class ListenableFutureCoroutine<T>(
...
+     * When the attached coroutine [isCompleted][Job.isCompleted] [exceptionally][Job.isCancelled]
+     * its outcome should be passed to this method.
+     *
+     * This method will map coroutine's exception into corresponding Future's exception.
+     *
+     * This should succeed barring a race with external cancellation.
+     */
+    // CancellationException is wrapped into `Result.Cancellation` to preserve original cause and message.
+    // All the other exceptions are delegated to SettableFuture.setException.
+    fun completeExceptionally(t: Throwable) =
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:366]: Please declare an explicit return type here for the sake of readability

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 5]
```diff
@@ -333,130 +329,161 @@ private class ListenableFutureCoroutine<T>(
...
+ *
+ * - When coroutine [isCompleted][Job.isCompleted] successfully, its result is wrapped into [Success].
+ * - When coroutine [isCompleted][Job.isCompleted] exceptionally, meaning [isCancelled]
+ *   returns `true`. According to [Future]'s contract, the exception can be of two types:
+ *     - [CancellationException] if the coroutine was _cancelled normally_, meaning it threw [CancellationException].
+ *       In this case the exception is wrapped into [Result.Cancellation].
+ *     - [ExecutionException] with the original [Throwable] in its [cause][ExecutionException.cause] otherwise.
+ *       This case is handled separately. See [JobListenableFuture.completeExceptionally].
+ */
+private sealed class Result<out T> {
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:481]: Thanks for the explanation!
Would suggest getting rid of the `Result` clas tho: `Result.success` means a different thing in Kotlin (name clash with a popular stdlib class is not a great thing to have) and we also have to allocate on successful completion.

We can use raw value for the successful results and a `Cancelled(exception)` wrapper for cancellation exception. `isCancelled` then can be replaced with `val Any.is[Future]Cancelled() get() = this is Cancelled`

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 6]
```diff
@@ -195,13 +196,21 @@ private fun ExecutionException.nonNullCause(): Throwable {
...
-    outerFuture.afterInit()
-    return outerFuture
+    val listenableFuture = JobListenableFuture<T>(this)
+    // This invokeOnCompletion completes the JobListenableFuture with the same result as `this` Deferred.
+    // The JobListenableFuture may have completed earlier if it got cancelled! See JobListenableFuture.cancel().
+    invokeOnCompletion { throwable ->
+        if (throwable == null) {
+            listenableFuture.complete(getCompleted())
+        } else {
+            listenableFuture.completeExceptionally(throwable)
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:210]: It seems like a bug if you don't know what's inside `completeExceptionally`. Please rename it to `completeExceptionallyOrCancel` or anything else that declares the intent

**Review by qwwdfsad — CHANGES_REQUESTED** [review]
Great job! A decent number of comments, every change and corner case is tested, really valuable contribution!

Let's polish it a bit and it's ready to go.

**Review by vadimsemenov — COMMENTED** [review]
Thanks for the review. I added another commit. I'll squash it into the previous one if it looks good to you.

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt (hunk 2 above)
Comment (qwwdfsad) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: You are indeed right, I've checked the wrong source

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 7]
```diff
@@ -152,7 +152,8 @@ public fun <T> ListenableFuture<T>.asDeferred(): Deferred<T> {
     deferred.invokeOnCompletion {
         cancel(false)
     }
-    return deferred
+    // Return hides the CompletableDeferred. This should prevent casting.
+    return object : Deferred<T> by deferred {}
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-guava/src/ListenableFuture.kt:156]: While we are here, please fix indent on line 170 (should be 4 spaces instead of 2)

**Review by qwwdfsad — APPROVED** [review]
Great! Please fix a minor comment and it's good to go.

I'm ready to squash & merge it and make a 1.4.2 release as soon as it's ready

File: integration/kotlinx-coroutines-guava/src/ListenableFuture.kt [hunk 8]
```diff
@@ -333,130 +329,154 @@ private class ListenableFutureCoroutine<T>(
...
 }
+
+/**
+ * A wrapper for `Coroutine`'s [CancellationException].
+ *
+ * If the coroutine is _cancelled normally_, we want to show the reason of cancellation to the user. Unfortunately,
+ * [SettableFuture] can't store the reason of cancellation. To mitigate this, we wrap cancellation exception into this
+ * class and pass it into [SettableFuture.complete]. See implementation of [JobListenableFuture].
+ */
+private class Cancelled(val exception: CancellationException)
```
[the remark this answers is not in the document]
Comment (vadimsemenov) [reply integration/kotlinx-coroutines-guava/src/ListenableFuture.kt]: Oops, the same mistake again, sorry.

**Comment by vadimsemenov** [comment]
Done.
I also filled #2409 to address the issue with two ignored tests when `AbstractFuture.afterDone` is finalised.

**Review by qwwdfsad — APPROVED** [review]
